### ISSUE TITLE
`T` should implement `From<NoneError>` where `T: Default`

### ISSUE THREAD

**Issue by Vurich** [rust-lang/rust#48951]
At the moment the documentation asks that users implement `From<NoneError>` manually for their error types, but this will rarely be more than a call to `Default::default()` and specialisation covers the cases where it is not.

**Comment by ExpHP** [comment]
The documentation reads:

> The error type that results from applying the try operator (`?`) to a `None` value. **If you wish to allow `x?` (where `x` is an `Option<T>`) to be converted into your error type,** you can implement `impl From<NoneError> for YourErrorType`.

Emphasis mine.  I do not believe this is desirable for most user Error types, even those that implement Default.

**Comment by Vurich** [comment]
Could you give an example of where it is undesirable? I think many functions (certainly many in my code) return a `Result<T, ZeroSizedErrorType>` for documentation purposes, which is isomorphic to `Option<T>`. The way I see it, if it was undesirable for a type to be converted from a `NoneError` then it would also be undesirable for that type to have a `Default` implementation.

**Comment by ExpHP** [comment]
It feels like there is footgun potential in the vicinity of `Option<Result<T, _>>` and `Result<Option<T>, _>` types.  For a contrived example:
```rust
let lines = BufReader::new(file).lines();

// (note: the author's intent here was to bail on Err.  Perhaps this was simply
//  done without much thought, by copying from the first line of the for loop)
let first = lines.next()?;

match first.split_whitespace().next() {
    ...
}

for line in lines {
    let line = line?;
    ...
}
```
when this gives a "no such method" error at `first.split_whitespace`, the user might add a `?` before `.split_whitespace()`, rather than noticing the real problem.

But as you say, the user's error type would need to impl `Default` for this to even be a problem.

---

I *was* going to say that it seemed probable that poorly written code contains general Error types with catastrophic derives like this:

```rust
#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
pub struct Error(String);
```

However, I tested this hypothesis by searching GitHub for [struct Error String](https://github.com/search?l=&p=5&q=struct+Error+String+language%3ARust&ref=advsearch&type=Code&utf8=✓) and could not find any supporting examples within the first few pages of results.

---

**Tl;dr:** I am as yet unable to come up with any concrete evidence of footgun potential.

**Comment by cramertj** [comment]
This seems the same as implementing `From<()>` for `T: Default`, and both are similar to `From<T> for ()`. Both are "correct" in some sense since there is only one obvious implementation, but I have trouble imagining where they'd be useful in practice, and they seem more likely to result in confusing code or inference failures.

**Comment by Vurich** [comment]
I don't think it's the same as `From<()> for T where T: Default`, which has an obvious footgun:

```rust
// Whoops, we changed the type signature between versions
fn function_that_returns_bar() {
}

let my_thing: Foo = function_that_returns_bar().into();
```

Implementing traits for `()` is different to implementing them for `NoneError`, since the former has a special meaning and is very common. You will rarely-if-ever even see `NoneError` referenced in your code, let alone have a function returning a bare `NoneError` like we have functions returning `()`.

A different, more explicit proposal would be `Option::ok_or_default()`, maybe that would be a nicer

**Comment by steveklabnik** [comment]
Triage: no changes

**Comment by xfix** [comment]
`NoneError` is gone, so this issue can be closed.
